diff --git a/userspace/libsinsp/CMakeLists.txt b/userspace/libsinsp/CMakeLists.txt
index ec454de7..4bac3120 100644
--- a/userspace/libsinsp/CMakeLists.txt
+++ b/userspace/libsinsp/CMakeLists.txt
@@ -33,6 +33,8 @@ if(NOT WIN32 AND NOT APPLE)
 	include_directories("${CMAKE_CURRENT_BINARY_DIR}")
 endif()
 
+include_directories("${PROJECT_SOURCE_DIR}/../falco-plugin")
+
 if(NOT WIN32)
 	include_directories("${CURL_INCLUDE_DIR}")
 endif()
@@ -116,6 +118,11 @@ set(SINSP_SOURCES
 	user_event.cpp
 	value_parser.cpp
 	viewinfo.cpp
+	../../../falco-plugin/tracer_interface.cpp
+	../../../falco-plugin/tracer.cpp	
+	../../../falco-plugin/tracer_common.cpp
+	../../../falco-plugin/tracer_msg_creator.cpp
+	../../../falco-plugin/tsc.cpp
 )
 
 if(WIN32)
diff --git a/userspace/libsinsp/gen_filter.cpp b/userspace/libsinsp/gen_filter.cpp
index 3b07a127..4e5b80c1 100644
--- a/userspace/libsinsp/gen_filter.cpp
+++ b/userspace/libsinsp/gen_filter.cpp
@@ -22,6 +22,10 @@ along with sysdig.  If not, see <http://www.gnu.org/licenses/>.
 #include "sinsp.h"
 #include "sinsp_int.h"
 
+#if TRACE_FALCO || TRACE_FALCO_RULES
+#include "../falco-plugin/tracer_interface.h"
+#endif
+
 gen_event::gen_event()
 {
 }
@@ -97,6 +101,11 @@ bool gen_event_filter_expression::compare(gen_event *evt)
 
 		if(j == 0)
 		{
+
+		#ifdef TRACE_FALCO_RULES
+			set_current_rule(chk->get_check_id());
+		#endif
+
 			switch(chk->m_boolop)
 			{
 			case BO_NONE:
